src/Data/Geometry/Geos/Raw/Serialize.hs

```diff
@@ -40,34 +42,40 @@ withWktWriter (WktWriter w) f = withForeignPtr w f
 
 createReader :: Geos Reader
 createReader = withGeos $ \h -> do
-    ptr <- throwIfNull "Create Reader" $
-      I.geos_WKBReaderCreate h
+    ptr <- throwIfNull "Create Reader" $ I.geos_WKBReaderCreate h
     fp <- newForeignPtrEnv I.geos_WKBReaderDestroy h ptr
     return $ Reader fp
 
 createWktReader :: Geos WktReader
 createWktReader = withGeos $ \h -> do
-    ptr <- throwIfNull "Create WKT Reader" $
-      I.geos_WKTReaderCreate h
+    ptr <- throwIfNull "Create WKT Reader" $ I.geos_WKTReaderCreate h
     fp <- newForeignPtrEnv I.geos_WKTReaderDestroy h ptr
     return $ WktReader fp
 
 read_ :: (I.GEOSContextHandle_t -> Ptr I.GEOSWKBReader -> CString  -> CSize -> IO (Ptr I.GEOSGeometry)) 
             -> Reader 
             -> BC.ByteString 
-            -> Geos Geom
-read_ f r bs = withGeos $ \h -> do
-  ptr <- withReader r $ \rp -> 
-            BC.useAsCStringLen bs $ \(cs, l) -> 
-              f h rp cs $ fromIntegral l 
-  fp <- newForeignPtrEnv I.geos_GeomDestroy h ptr
-  return $ Geom fp
-  
-
-read :: Reader -> BC.ByteString -> Geos Geom
+            -> Geos (Maybe Geom)
+read_ f r bs = withGeos $ \h -> 
+    onException (readBlock h) (return Nothing)
+      where
+        readBlock h =  do
+            mptr <- withReader r $ \rp -> 
+               BC.useAsCStringLen bs $ \(cs, l) -> do
+                        p <- f h rp cs $ fromIntegral l
+                        if p == nullPtr
```

Comment:
Took me way too long to realize you could just compare with nullPtr to handle this.